README.md
# plotMap versão 0.132
O projeto 'plotMap' tem por objetivo fornecer uma forma simples de plotar mapas.
O código é livre (licença MIT) e conto com a colaboração de todos para o seu aperfeiçoamento.

<img src="mapa_home.png" width="200"> <img src="mapa_home_2.png" width="200">

## Utilização:

Em breve. 
Mas o arquivo 'exemplo.py' já lhe dará uma boa noção de como utilizar o código.

## Dependências:

Matplotlib e Cartopy.

## Licença:

[Ver licença](LICENSE)


## Projeto relacionado:

[mapLoaders](https://github.com/NelsonBittencourt/mapLoaders)


## Sobre o autor:

[Meu LinkedIn](http://www.linkedin.com/in/nelsonrossibittencourt)

[Minha página de projetos](http://www.nrbenergia.somee.com)




changes.txt
Log de Inserção/alteração de arquivos
-------------------------------------

Em 13/05/2021:
	- Arquivos de template atualizados para a versão 0.111, que inclui a chave 'mapa_tipo'.
	Está nova chave permite que se escolha entre 'contornos' e 'xy' (scatter plot) como tipo de mapa.

Em 06/05/2021:
    - Arquivos de template atualizados para a versão 0.11, que incluí os parâmetros de tamanho e posição relativa da barra de cores.

Em 05/05/2021:
    - Inserido este arquivo de log;
    - Inserido o arquivo 'templates/AnomaliaWxmaps.dat' com configurações para mapa de anomalias do site WXMaps.


Log de Alterações da Rotina 'plotMap' (arquivo 'plotMap.py')
------------------------------------------------------------

Versão 0.132 (23/05/2021):
-Adição do tipo de mapa 'scatter plot' ou xy. As seguintes funções foram alteradas:	'plotarMapa e loadMapTemplate'

Versão 0.131 (18/05/2021):
- Adição de múltiplos arquivos shape. Foram necessárias alterações na definição da classe 'ArquivoShape' e na Função
'plotarMapa'.


Versão 0.13 (06/05/2021):
- Classe 'Mapa', Função 'plotarMapa' e Função 'loadMapTemplate':
    Adicionados os parâmetros de ajuste de tamanho e posição relativa da barra de cores.

- Função 'plotarMapa': 
    Limpeza da sintaxe do comando necessário para criar a barra de cores.

Versão 0.12 (05/05/2021):

- Função 'plotarMapa':
    Correção de bug no método 'contourf'. Obrigatório declarar o número de níveis.

Versão 0.11 (05/05/2021):

- Adicionada a classe 'shapeFile';

- Função 'plotarMapa':
    O parâmetro 'shapeFile' é agora opcional;
    Atualização da sintaxe dos métodos do 'matplotlib' para a versão 3.4;
    Simplificação na rotina no ajuste do mapa de cores;
    Ajustes na rotina de barra de cores.

- Função 'loadMapTemplate':
    Adicionado tratamento de erros na leitura do arquivo;
    Adicionado tratamento de erros na alocação de variáveis no objeto 'Mapa'.




Log de Alterações do arquivo de exemplos ('exemplos.py'):
---------------------------------------------------------

Versão 0.121 (13/05/2021):
- Adição de múltiplos arquivos shape. Foram necessárias alterações nos exemplos.

Versão 0.12 (06/05/2021):
- Alteração no exemplo 2 para utilizar o parâmetro de distância da barra de cores.

Versão 0.11 (05/05/2021):
- Simplificação na leitura do arquivo binário de exemplo;
- Inserção do exemplo 4, que utiliza o modelo 'AnomaliaWxmaps.dat';
- Utilização do objeto 'ArquivoShape' para evitar leituras desnecessárias (e lentas) do arquivo '.shp' dos três primeiros exemplos.









test.py
# -*- coding: utf-8 -*-

import plotMap
import numpy as np
import struct

chuva = []

arquivo_dados = open('ETA_24_000.bin','rb')
    # Loop para ler os dados do grads
try:
    byte1 = arquivo_dados.read(4)                           # lê 4 bytes 
    chuva.append(struct.unpack('f', byte1))                 # Converte para float
    while byte1 != b"":                                     # Repete o processo até o fim do arquivo
        byte1 = arquivo_dados.read(4)
        if len(byte1)==4:                                   # Verifica se o dado lido é mesmo um byte
            chuva.append(struct.unpack('f', byte1))        
finally:
    arquivo_dados.close()                                   # Fecha o arquivo em caso de erro


lons = np.arange(-83.0,-25.6,0.4)
lats = np.arange(-50.2,12.6,0.4)
chuva = np.array(chuva)
chuva = np.reshape(chuva,(157,-1))
shapeFile = 'BRA_adm1.shp'
mapTemplate = 'mapateste.dat'
titulo_mapa = 'Precipitação (mm)\nTeste'

# Teste 1: usando o template do mapa
plotMap.plotarMapa(titulo_mapa, lons,lats,chuva,shapeFile,mapTemplate)

# Teste 2: carregando o template e alterando uma característica
meuMapaCustomizado = plotMap.loadMapTemplate(mapTemplate)
meuMapaCustomizado.barraCores_titulo = "barra de cores alterada!"
meuMapaCustomizado.barraCores_orientacao = 'vertical'
plotMap.plotarMapa(titulo_mapa, lons,lats,chuva,shapeFile,meuMapaCustomizado, 'Teste2.jpg')


plotMap.py
# -*- coding: utf-8 -*-

"""
******************************************************************************
plotMap.py - Módulo para auxiliar na plotagem de mapas com dados

Autor   : Nelson Rossi Bittencourt
Versão  : 0.132
Licença : MIT
Dependências: matplotlib e cartopy
******************************************************************************
"""

import matplotlib as mpl
import matplotlib.pyplot as plt
import cartopy
import cartopy.crs as ccrs
from cartopy.io.shapereader import Reader
from cartopy.feature import ShapelyFeature

# TODO: implementar as demais características do mapa.


# Classes

class Mapa:    
    """
    Classe Mapa - Define o layout do mapa a plotar.

    É utilizado nas rotinas 'plotarMapa' e 'loadMapTemplate'.

    Também poderá ser instanciada pelo usuário para criar modelos de mapas próprios ou modelos de mapas lidos de arquivos.
    
    """
    def __init__(self):
        #self.mapa_nome = ''
        #self.mapa_tipo = '' 
        self.mapa_coordenadas = []       
        self.mapa_tipo = ''
        self.barraCores_titulo = ''        
        self.barraCores_orientacao = ''  
        self.barraCores_valores = []  
        self.barraCores_codigos = []
        self.barraCores_posicao = ''
        self.barraCores_corMinimo = ''
        self.barraCores_corMaximo = ''
        self.barraCores_dist=0
        self.barraCores_tam=0        

class ArquivoShape:
    """
    Classe ArquivoShape - Representa as características de um arquivo tipo 'shape file' para inserção em um mapa.

    Para criar uma nova instância dessa classe é preciso fornecer o nome completo do arquivo 'shp'.
    
    Esta classe permite que um mesmo arquivo shape seja utilizado em diversos mapas, sem o overhead causado
    pela leitura do arquivo diversas vezes.

    """
    def __init__(self,nomeArquivo, corFace='none', corLinha='gray',espLinha=0.5):        
        self.shape_feature =    ShapelyFeature(Reader(nomeArquivo).geometries(), ccrs.PlateCarree(),
                                facecolor=corFace, edgecolor=corLinha, linewidth=espLinha,) 
                  


# Funções

def plotarMapa(titulo, lons, lats, dados, modeloMapa, destino='', shapeFile=-1):
    """
    Plota um mapa considerando os dados fornecidos.
    
    Argumentos
    ----------
    titulo : Título do mapa;
    
    lons : Lista com as longitutes;
    
    lats : Lista com as latitudes;
    
    dados : Lista com os dados a plotar;        
    
    modeloMapa  : string ou objeto tipo 'Mapa'. 
        A string deve conter um nome de arquivo de template válido.
        Mapa deve conter uma instância do tipo 'Mapa' válida.
    
    destino : (Opcional) Nome do arquivo de saída para a figura. Se não declarado, exibe mapa na tela.
    
    shapeFile   : (Opcional) Deve ser fornecido:
        O nome do arquivo tipo 'shp' para leitura do disco

        ou um objeto tipo 'ArquivoShape' instanciado previamente

        ou uma lista contendo uma combinação de strings e objetos tipo 'ArquivoShape'
    
    Retorno
    -------    
        Nenhum.    
    """

    # Verifica o tipo de argumento passado em 'modeloMapa'.

    tipoModelo = type(modeloMapa)
    if tipoModelo is str:
        myMap = loadMapTemplate(modeloMapa)
    elif tipoModelo is Mapa:
        myMap = modeloMapa
    else:
        raise NameError("O argumento 'modeloMapa' deve ser uma string ou um tipo 'Mapa'!")
    
    # Fecha uma figura anterior, se houver.    
    plt.close()

    # Determina o tamanho do gráfico no console do Python.
    fig = plt.figure(figsize=(5,5))  

    # Detemina tipo de projeção.
    ax = fig.add_subplot(1,1,1,projection=ccrs.PlateCarree())
   
    # Delimita o mapa.
    ax.set_extent(myMap.mapa_coordenadas,ccrs.PlateCarree())
    
    # Adiciona algumas caracteristica no mapa.
    # Talvez, no primeiro uso, o 'matplotlib'/'cartopy' execute o download de mapas com as características requeridas.
    ax.add_feature(cartopy.feature.LAND)   
    ax.add_feature(cartopy.feature.COASTLINE)
    ax.add_feature(cartopy.feature.BORDERS)
    
    # Adiciona arquivos tipo 'shape' ao mapa.
    # Lista que conterá (ou não) strings, 'ArquivosShape' ou uma combinação.
    newList = []

    # Caso 'shapeFile' não seja uma lista, cria uma lista.
    if (type(shapeFile)) is not list:
        newList.append(shapeFile)
    else:
        newList = shapeFile

    # Para cada item da lista 'newList', verifica o tipo e adiciona a característica no mapa
    for item in newList:
        tipoItem = type(item)
        if tipoItem is ArquivoShape:
            ax.add_feature(item.shape_feature)
        elif tipoItem is str:
            tmp = ArquivoShape(item)
            ax.add_feature(tmp.shape_feature)
           
    # Variáveis auxiliares para ajuste do mapa de cores
    infbound = None
    supbound = None
    extend = 'neither'

    # Ajusta código de cores se necessário.
    if (myMap.barraCores_corMinimo!='-1') and (myMap.barraCores_corMaximo!='-1'): 
        extend = 'both'
        infbound = myMap.barraCores_corMinimo
        supbound = myMap.barraCores_corMaximo
    elif (myMap.barraCores_corMinimo!='-1'): 
        extend = 'min'
        infbound = myMap.barraCores_corMinimo
    else: 
        extend = 'max'
        supbound = myMap.barraCores_corMaximo

    # Cria mapa de cores.        
    cmap = (mpl.colors.ListedColormap(myMap.barraCores_codigos).with_extremes(over=supbound, under=infbound))
    
    # Cria o índice de cores do mapa
    norm = mpl.colors.BoundaryNorm(myMap.barraCores_valores, cmap.N)

    # Cria o gráfico de acordo com o tipo selecionado.
    if (myMap.mapa_tipo=='contornos'):
        filled = ax.contourf(lons, lats, dados, levels=myMap.barraCores_valores,cmap=cmap, norm=norm, extend=extend, transform=ccrs.PlateCarree())
    elif (myMap.mapa_tipo=='xy'):
        filled = ax.scatter(lons, lats, c=dados, s=50,alpha=1,cmap=cmap, norm=norm,edgecolors='black', transform=ccrs.PlateCarree())
    else:
        raise NameError("Tipo de mapa (mapa_tipo) inválido! Verifique o arquivo de template.")

    # Ajusta a barra de cores, se houver.    
    if myMap.barraCores_orientacao!="none":       

        cbar = fig.colorbar(
                    filled,
                    orientation= myMap.barraCores_orientacao,
                    label= myMap.barraCores_titulo,
                    spacing = 'uniform',
                    pad=myMap.barraCores_dist,
                    fraction=myMap.barraCores_tam,                      
                    location=myMap.barraCores_posicao,                    
                    extendfrac='auto',                                       
                    )

        cbar.set_ticks(myMap.barraCores_valores)        
    
    # Define título do gráfico.
    plt.title(titulo)
  
    # Adiciona grid.         
    g1=ax.gridlines(crs=ccrs.PlateCarree(), draw_labels=True,linewidth=1, color='gray', alpha=0.5, linestyle='--')
    g1.top_labels = False
    g1.right_labels = False
    g1.xlabel_style = {'size': 9, 'color': 'blue', 'weight': 'bold'}
    g1.ylabel_style = {'size': 9, 'color': 'red', 'weight': 'bold'}
    
    # Mostra na tela ou salva em arquivo.
    if destino == '':        
        plt.show(block=True)
        
    else:              
        fig.savefig(destino)
    

def loadMapTemplate(arquivoTemplateMapa):
    """
    Lê as características de um modelo de mapa a partir de um arquivo de texto formatado.
    
    Argumentos
    ----------

    arquivoTemplateMapa : nome do arquivo contendo template do Mapa.
        Este arquivo deve seguir, estritamente, o formato estabelecido.

    Retorno
    -------
    objeto Mapa (ver definição da classe 'Mapa' neste arquivo).        
    
    """
    
    # Lista com as linhas lidas do arquivo de template de um mapa.
    lines = []

    # Número de linhas válidas esperado.
    check_valid_lines = 11

    # Contador de linhas válidas.
    valid_lines = 0

    # Número da linha atual que está sendo lida
    num_line = 0

    # Dicionário com dados lidos.
    map_dict = {}

    # Abre arquivo e lê as linhas válidas.
    # O caracter '#' é utilizado para comentários no arquivo de template.
    # Se o número de linhas válidas for diferente do esperado, lança uma exceção. 
    try:

        with open(arquivoTemplateMapa, 'r') as f:

            for line in f:
                line = line.rstrip() 
                num_line = num_line + 1          
                prefix = line.strip()

                if len(prefix) > 0 and prefix[0] != '#':
                    if (':' in line):
                        lines.append(line)
                        valid_lines = valid_lines + 1
                        listValues = line.split(':')                    
                        map_dict[listValues[0]]= listValues[1]                    
                    else:
                        raise NameError("Linha inválida no arquivo de template '{}'. Verique a linha {}.".format(arquivoTemplateMapa, num_line))
                        
    except:
        raise NameError("Erro ao tentar abrir o arquivo de template para o mapa [{}]!\nVerifique o caminho completo do arquivo e tente novamente.".format(arquivoTemplateMapa))        
   

    # Verifica se o arquivo contem o número de linhas válidas.
    if valid_lines!=check_valid_lines:
        raise NameError("O arquivo de template '{}' contêm {} linhas válidas, quando o esperado são {} linhas.".format(arquivoTemplateMapa,valid_lines,check_valid_lines))


    # Aloca os valores em um objeto do tipo 'Mapa'
    # TODO: Comentar alocações abaixo, se necessário
    local_map = Mapa()        
    
    try:
        local_map.mapa_tipo = map_dict['mapa_tipo']

        local_map.barraCores_orientacao = map_dict['barra_cores_orientacao']
        local_map.barraCores_titulo = map_dict['barra_cores_titulo'] 

        local_map.barraCores_corMinimo = map_dict['barra_cores_corMinimo']
        local_map.barraCores_corMaximo = map_dict['barra_cores_corMaximo']

        local_map.barraCores_posicao = map_dict['barra_cores_posicao']
        local_map.barraCores_codigos = map_dict['barra_cores_codigos'].split(',')   

        local_map.barraCores_dist = float(map_dict['barra_cores_dist'])
        local_map.barraCores_tam = float(map_dict['barra_cores_tam'])         

        tmp = map_dict['barra_cores_valores'].split(',')    
        local_map.barraCores_valores = [float(i) for i in tmp]    

        tmp = map_dict['mapa_coordenadas'].split(',')   
        local_map.mapa_coordenadas = [float(i) for i in tmp]

    except:
        raise NameError("Erro ao tentar interpretar o arquivo de template [{}] para o mapa!\nVerifique a sintaxe do arquivo e tente novamente.".format(arquivoTemplateMapa))
           
    return(local_map)
        



PyMaps_Cartopy.py
# -*- coding: utf-8 -*-
"""
Created on Sat Aug 25 20:35:13 2018

@author: Nelson Rossi Bittencourt

******************************************************************************
Rotina para plotar dados de chuva a partir de uma arquivo no formato GRADS 
Neste exemplo, o arquivo GRADS tem nome 'ETA_24_000.bin'
O arquivo 'shad50_1.gif' é uma representação oficial do arquivo 'bin' acima
Esta rotina é somente uma demonstração e não foi otimizada para uso comercial
******************************************************************************

Alterado em 19/03/2021 para ser executado no Python 3.7 (VS Code)

"""

# Importação das dependências
# Duas linhas abaixo foram removidas em 19/03/2021. Eram utilizada na versão anterior do Python
# import os
# os.environ["PROJ_LIB"] = "C:\\ProgramData\\Anaconda3\\Library\\share"

# Em 26/03/2021 
# Foi necessário instalar o matplotlib no Windows com: py -m pip install matplotlib
# Foi necessário baixar e instalar 'basemap-1.2.2-cp39-cp39-win_amd64.whl'
#   O arquivo foi obtido em https://www.lfd.uci.edu/~gohlke/pythonlibs/
#   O arquivo foi copiado no diretório atual e instalado com py -m pip install basemap-1.2.2-cp39-cp39-win_amd64.whl

# Em 20/04/2021
# Instalar Cartopy do mesmo site acima
# Instalar scipy


import matplotlib as mpl
import matplotlib.pyplot as plt
import cartopy
import cartopy.crs as ccrs
from cartopy.io.shapereader import Reader
from cartopy.feature import ShapelyFeature

import numpy as np
import struct

#import plotMap
#from plotMap import Mapa

#meuMapa1 = Mapa()
#meuMapa2 = Mapa()
#meuMapa1.nome = 'Teste1'
#meuMapa2.nome = 'Teste2'
#plotMap.testeMapa(meuMapa2)




# Abrir arquivo de dados binários com dados de chuva
# O arquivo está no formato GRADS sendo necessário convertê-lo
arquivo_dados = open('ETA_24_000.bin','rb')

# Lista temporária contendos os dados de chuva
chuva = []

# Loop para ler os dados do grads
try:
    byte1 = arquivo_dados.read(4)                           # lê 4 bytes 
    chuva.append(struct.unpack('f', byte1)[0])              # Converte para float
    while byte1 != b"":                                     # Repete o processo até o fim do arquivo
        byte1 = arquivo_dados.read(4)
        if len(byte1)==4:                                   # Verifica se o dado lido é mesmo um byte
            chuva.append(struct.unpack('f', byte1)[0])       
finally:
    arquivo_dados.close()                                   # Fecha o arquivo em caso de erro
                                                    

# Fecha o arquivo de dados em condição normal
arquivo_dados.close()

# Converte a lista em matriz e redimensiona
# Este procedimento é necessário para adequar as dimensões das matrizes 
# no momento de utilizar a função p
chuva = np.array(chuva)
chuva = np.reshape(chuva,(157,-1))

# longitudes
lons = np.arange(-83.0,-25.6,0.4)

# latitudes
lats = np.arange(-50.2,12.6,0.4)

# Determina o tamanho do gráfico no console do Python
fig = plt.figure(figsize=(5,5))

# Detemina tipo de projeção
ax = fig.add_subplot(1,1,1,projection=ccrs.PlateCarree())

# Determina coordenada do mapa

# set_extent(loni, lonf, lati, latf)
ax.set_extent([-75,-35,-35,5],ccrs.PlateCarree() )



# Adiciona as caracteristica do mapa
ax.add_feature(cartopy.feature.LAND)
#ax.add_feature(cartopy.feature.OCEAN)
ax.add_feature(cartopy.feature.COASTLINE)
ax.add_feature(cartopy.feature.BORDERS)


fname = "BRA_adm1.shp"
shape_feature = ShapelyFeature(Reader(fname).geometries(), ccrs.PlateCarree(), facecolor='none')

ax.add_feature(shape_feature)

# Cria a escala de valores para escala de cores
valores_escala = [0,5,10,15,20,25,30,40,50,75,100,150,200]
etiquetas_escala = valores_escala

# Cria o gráfico do tipo contornos preenchidos
filled = ax.contourf(lons, lats, chuva, transform=ccrs.PlateCarree(), cmap='hsv')

# Ajusta a barra de cores
cmap = mpl.cm.viridis.with_extremes(over='red', under='white')
norm = mpl.colors.BoundaryNorm(valores_escala, cmap.N, extend='both')

cbar = fig.colorbar(
            mpl.cm.ScalarMappable(norm=norm, cmap=cmap),
            orientation='horizontal',
            label="Precipitação (mm)",
            spacing = 'uniform')


cbar.set_ticks(valores_escala)
cbar.set_ticklabels(etiquetas_escala)

# Carrega um arquivo do tipo shapefile    
# Foram mantidos todos os arquivos 'BRA_adm1' por convenção.
# TODO: verificar que arquivos 'BRA_adm1' são realmente necessários
#m.readshapefile('BRA_adm1', 'Brasil')

# Define título do gráfico 
plt.title('Chuva Prevista - Modelo ETA (mm)')

# Adiciona grid de coordenadas ao mapa
#paralelos = np.arange(-35,5,5)
#meridianos = np.arange(-75,-30,5)
#m.drawparallels(paralelos,labels=[True,False,False,False])
#m.drawmeridians(meridianos,labels=[False,False,False,True])

# Salva o gráfico em um arquivo png com resolução de 200 dpi
#plt.savefig('arquivo_saida.png', bbox_inches='tight', dpi=240)
plt.show()  # Adicionado em 19/03/2021


PyMaps_Inicial.py
# -*- coding: utf-8 -*-
"""
Created on Sat Aug 25 20:35:13 2018

@author: Nelson Rossi Bittencourt

******************************************************************************
Rotina para plotar dados de chuva a partir de uma arquivo no formato GRADS 
Neste exemplo, o arquivo GRADS tem nome 'ETA_24_000.bin'
O arquivo 'shad50_1.gif' é uma representação oficial do arquivo 'bin' acima
Esta rotina é somente uma demonstração e não foi otimizada para uso comercial
******************************************************************************

Alterado em 19/03/2021 para ser executado no Python 3.7 (VS Code)

"""

# Importação das dependências
# Duas linhas abaixo foram removidas em 19/03/2021. Eram utilizada na versão anterior do Python
# import os
# os.environ["PROJ_LIB"] = "C:\\ProgramData\\Anaconda3\\Library\\share"

# Em 26/03/2021 
# Foi necessário instalar o matplotlib no Windows com: py -m pip install matplotlib
# Foi necessário baixar e instalar 'basemap-1.2.2-cp39-cp39-win_amd64.whl'
#   O arquivo foi obtido em https://www.lfd.uci.edu/~gohlke/pythonlibs/
#   O arquivo foi copiado no diretório atual e instalado com py -m pip install basemap-1.2.2-cp39-cp39-win_amd64.whl

# Em 15/05/2020
# Teste sem 'basemap'
#import plotly.express as px
#import pandas as pd

import matplotlib.pyplot as plt
#import plotly.express as px



from mpl_toolkits.basemap import Basemap, cm
import numpy as np
import struct

# Abrir arquivo de dados binários com dados de chuva
# O arquivo está no formato GRADS sendo necessário convertê-lo
arquivo_dados = open('ETA_24_000.bin','rb')

# Lista temporária contendos os dados de chuva
chuva = []

# Loop para ler os dados do grads
try:
    byte1 = arquivo_dados.read(4)                           # lê 4 bytes 
    chuva.append(struct.unpack('f', byte1)[0])              # Converte para float
    while byte1 != b"":                                     # Repete o processo até o fim do arquivo
        byte1 = arquivo_dados.read(4)
        if len(byte1)==4:                                   # Verifica se o dado lido é mesmo um byte
            chuva.append(struct.unpack('f', byte1)[0])       
finally:
    arquivo_dados.close()                                   # Fecha o arquivo em caso de erro
                                                    

# Fecha o arquivo de dados em condição normal
arquivo_dados.close()

# Converte a lista em matriz e redimensiona
# Este procedimento é necessário para adequar as dimensões das matrizes 
# no momento de utilizar a função p
chuva = np.array(chuva)
chuva = np.reshape(chuva,(157,-1))

# longitudes
lons = np.arange(-83.0,-25.6,0.4)

# latitudes
lats = np.arange(-50.2,12.6,0.4)

# Determina o tamanho do gráfico no console do Python
fig = plt.figure(figsize=(5,10))

# Determina os eixos do gráfico
ax = fig.add_axes([0,0,1,1])

# Define as característica do mapa
# A resolução mais baixa é 'c' e a mais alta é 'f'
# Para resoluções altas, pode ser necessário instalar pacotes adicionais
m = Basemap(resolution='c', # c, l, i, h, f or None
            projection='merc',
            lat_0=-15, lon_0=-52.5,
            llcrnrlon=-75., llcrnrlat= -35, urcrnrlon=-35, urcrnrlat=5)


# Criar um grid retangular com as coordenadas dadas
xi,yi = np.meshgrid(lons,lats)

# Converte o grid retangular para pontos de grade do mapa
# Essa conversão pode ser feita pela opçõa latlon = 'True' do comando contourf
# x, y = m(xi,yi)

# Cria a escala de valores para escala de cores
valores_escala = [0,1,5,10,15,20,25,30,40,50,75,100,150,200,999]

# Cria o gráfico do tipo contornos preenchidos
cs = m.contourf(xi,yi,chuva,valores_escala,cmap = cm.s3pcpn,latlon='True')

# Ajusta a barra de cores
cbar = m.colorbar(cs,location='bottom',pad="7%")
cbar.set_ticks([0,1,5,10,15,20,25,30,40,50,75,100,150,300,999])
cbar.set_ticks(valores_escala)
cbar.set_ticklabels(['   Mín',1,5,10,15,20,25,30,40,50,75,100,150,200,'  Máx'])

# Carrega um arquivo do tipo shapefile    
# Foram mantidos todos os arquivos 'BRA_adm1' por convenção.
# TODO: verificar que arquivos 'BRA_adm1' são realmente necessários
m.readshapefile('BRA_adm1', 'Brasil')

# Define título do gráfico 
plt.title('Chuva Prevista - Modelo ETA (mm)')

# Adiciona grid de coordenadas ao mapa
paralelos = np.arange(-35,5,5)
meridianos = np.arange(-75,-30,5)
m.drawparallels(paralelos,labels=[True,False,False,False])
m.drawmeridians(meridianos,labels=[False,False,False,True])

# Salva o gráfico em um arquivo png com resolução de 200 dpi
#plt.savefig('arquivo_saida.png', bbox_inches='tight', dpi=240)
#plt.show()  # Adicionado em 19/03/2021


df = pd.read_csv('https://raw.githubusercontent.com/plotly/datasets/master/earthquakes-23k.csv')

fig = px.density_mapbox(df, lat='Latitude', lon='Longitude', z='Magnitude', radius=10,
                        center=dict(lat=0, lon=180), zoom=0,
                        mapbox_style="stamen-terrain")
fig.show()


#fig = px.density_mapbox(chuva, lat = lats, lon = lons, z = chuva, radius= 1)
#fig.show()


exemplos.py
# -*- coding: utf-8 -*-

"""
******************************************************************************
Rotina para testar o módulo 'plotMap'
Nestes teste será lido um arquivo binário no formato 'Grads' com dados de chuva
prevista. Os dados deste arquivo serão plotados em um mapa com modelo semelhante
ao utilizado pelo ONS.

Autor   : Nelson Rossi Bittencourt
Versão  : 0.121
Licença : MIT
Dependências: numpy, struct, plotMap
******************************************************************************
"""

import numpy as np
import struct
import plotMap

def main():

    # Lista para acomodar os dadoss de chuva a serem lidos do arquivo tipo 'Grads'.
    chuva = []
  
    # Abre o arquivo 'ETA_24_000.bin' e aloca seus dados na lista 'chuva'.
    arquivo_dados = 'ETA_24_000.bin'
    
    # Lê o arquivo de dados binário tipo 'Grads'    
    try:
        with open(arquivo_dados, 'rb') as f:
            while True:
                byte = f.read(4)
                if not byte:
                    break                
                chuva.append(struct.unpack('f',byte))                   
    except:
        raise NameError('Erro ao tentar abrir/acessar arquivo: {}'.format('ETA_24_000.bin'))       
  

    # listas com as coordenads (longitudes e latitudes) do arquivo 'Grads'.
    # Estes valores poderiam ser lidos do arquivo de configuração. Por simplicidade, foram inseridos no código.
    # Ver método 'arange' do 'numpy' para detalhes do funcionamento da construção das listas.
    lons = np.arange(-83.0,-25.6,0.4)
    lats = np.arange(-50.2,12.6,0.4)

    # Transforma a lista 'chuva' em uma matriz e altera o seu 'formato' para compatibilidade com a o número de longitutes 
    # e latitudes. Veja o método 'reshape' do 'numpy' para maiores detalhes.
    chuva = np.array(chuva,dtype=float)    
    chuva = np.reshape(chuva,(157,-1))
    
    # Nome do arquivo com parâmetros do mapa.
    mapTemplate = 'templates/ChuvaPrevistaONS.dat'

    # Título do mapa
    titulo_mapa = 'Precipitação (mm)\nTeste'
        
    # Utilize a classe 'ArquivoShape' para criar uma instância de um arquivo shape com uma cor e uma 
    # espessura da linha especificados.
    # Para utilizar arquivos tipo 'shape', são necessários os arquivos '*.shp', '*.shx' e '*.dbf'.
    # Repare que você ainda poderá utilizar como 'shapeFile' apenas uma 'string' ou um objeto 'ArquivoShape'.

    # Criando um objeto 'ArquivoShape' com os Estados do Brasil, cor da linha cinza e espessura da linha 1.0.
    shapeEstadosBrasil = plotMap.ArquivoShape(nomeArquivo='shapes/BRA_adm1.shp', corLinha='gray', espLinha=1.0)
    
    # Criando um objeto 'ArquivoShape' com das bacias hidrográficas , cor da linha azul e espessura da linha 0.25.
    shapeBaciasBrasil =  plotMap.ArquivoShape(nomeArquivo='shapes/Bacias.shp', corLinha='blue', espLinha=0.25)    
    
    # Lista contendo os dois 'ArquivosShape' acima. Ao passar uma lista de shapes, a rotina 'plotarMapa' irá adicionar
    # cada arquivo ao mapa na mesma ordem da lista.
    listaShapes = [shapeEstadosBrasil, shapeBaciasBrasil]


    # Exemplo 1: Usando o modelo de mapa de forma integral.
    # Neste caso o mapa utilizará todos os parâmetros do arquivo especificado em 'mapTemplate'.
    # Também usaremos os 'shapes' dos Estados e das bacias.         
    plotMap.plotarMapa( titulo=titulo_mapa,
                        lons=lons,
                        lats=lats, 
                        dados=chuva,
                        modeloMapa=mapTemplate,                       
                        destino='exemplo1_mapaTipoONS.jpg',
                        shapeFile=listaShapes                   
                        )


    # Exemplo 2: Usando o modelo de mapa de forma parcial.
    # Neste exemplo vamos ler os parâmetros do mapa modelo para o objeto 'Mapa' e alterar alguns.
    # O método 'loadMapTemplate' recebe o nome do arquivo de modelo do mapa e retorna um objeto 'Mapa'. 
    # Usaremos somente o 'shape' dos Estados.   
    meuMapaCustomizado = plotMap.loadMapTemplate(mapTemplate)

    # Com o objeto 'Mapa', podemos alterar quaisquer dos parâmetros.
    meuMapaCustomizado.barraCores_titulo = "barra de cores alterada!"
    meuMapaCustomizado.barraCores_orientacao = 'vertical'
    meuMapaCustomizado.barraCores_posicao = 'right' 
    meuMapaCustomizado.barraCores_dist = 0.05
    plotMap.plotarMapa(titulo_mapa, lons,lats,chuva,meuMapaCustomizado, 'exemplo2_MapaCustomizado.jpg',listaShapes[0])


    # Exemplo 3: Repetindo o exemplo 2, com mapa na tela.
    # Para isso, basta não informar nada como 'destino' (sexto argumento) do método 'plotarMapa'.
    # Importante: como o parâmetro 'destino' não foi definido, é mandatório nomear o parâmetro posterior a ele na
    # lista de argumentos (neste caso 'shapeFile').
    # Usaremos somente o 'shape' das bacias.
    plotMap.plotarMapa(titulo_mapa, lons,lats,chuva, meuMapaCustomizado, shapeFile=listaShapes[1])


    # Exemplo 4: Repetindo o exemplo 2, com mapa do tipo 'anomalias' do site WXMaps.
    # Neste exemplo, não será utilizado 'shapefile'.   
    plotMap.plotarMapa( titulo='Anomalia (mm)\nWXMaps',
                        lons=lons,
                        lats=lats,
                        dados=chuva,
                        modeloMapa='templates/AnomaliaWxmaps.dat',
                        destino='exemplo4_MapaTipoWXMaps_Anomalia.jpg'
                       )    


if __name__ == '__main__':
    main()